src/services/codefixes/inferFromUsage.ts

```diff
@@ -275,11 +303,43 @@ namespace ts.codefix {
                 const typeExpression = createJSDocTypeExpression(typeNode);
                 const typeTag = isGetAccessorDeclaration(declaration) ? createJSDocReturnTag(typeExpression, "") : createJSDocTypeTag(typeExpression, "");
                 addJSDocTags(changes, sourceFile, parent, [typeTag]);
+                return;
             }
-            else {
-                changes.tryInsertTypeAnnotation(sourceFile, declaration, typeNode);
+            if (tryReplaceImportTypeNodeWithAutoImport(typeNode, changes, sourceFile, declaration, type, program, host, formatContext, preferences)) {
+                return;
+            }
+            changes.tryInsertTypeAnnotation(sourceFile, declaration, typeNode);
+        }
+    }
+
+    function tryReplaceImportTypeNodeWithAutoImport(typeNode: TypeNode, changes: textChanges.ChangeTracker, sourceFile: SourceFile, declaration: textChanges.TypeAnnotatable, type: Type, program: Program, host: LanguageServiceHost, formatContext: formatting.FormatContext, preferences: UserPreferences): boolean {
+        if (isLiteralImportTypeNode(typeNode) && typeNode.qualifier && type.symbol) {
+            // Replace 'import("./a").SomeType' with 'SomeType' and an actual import if possible
+            const moduleSymbol = find(type.symbol.declarations, d => !!d.getSourceFile().externalModuleIndicator)?.getSourceFile().symbol;
+            // Symbol for the left-most thing after the dot
+            const symbol = getLeftMostIdentifierOfEntityName(typeNode.qualifier).symbol;
+            if (moduleSymbol) {
+                const action = getImportCompletionAction(
+                    symbol,
+                    moduleSymbol,
+                    sourceFile,
+                    symbol.name,
+                    host,
+                    program,
+                    formatContext,
+                    declaration.pos,
+                    preferences,
+                );
+                if (action.codeAction.changes.length && changes.tryInsertTypeAnnotation(sourceFile, declaration, createTypeReferenceNode(typeNode.qualifier, typeNode.typeArguments))) {
+                    for (const change of action.codeAction.changes) {
+                        const file = sourceFile.fileName === change.fileName ? sourceFile : Debug.assertDefined(program.getSourceFile(change.fileName));
+                        changes.pushRaw(file, change);

```

Comment:
Is this the first codefix to run another codefix and merge its actions? Seems like something that would have a common utility inside the codefox folder.

Reply:
Completions calls functions from importFixes, but AFAIK no other codefix has called another codefix.